### ISSUE TITLE
Pkg.test("Circuitscape") fail in double precision, raster pairwise and advanced (Ubuntu)

### ISSUE THREAD

**Issue by EveTC** [Circuitscape/Circuitscape.jl#302]
Hello I was wondering if you could help me, I am hoping to use `julia circuitscape` within the R package `ResistanceGA`. However, I am getting an error during installation of circuitscape in Julia.

I am very new to the language Julia so any help would be grealty appreciated.

I have recently installed julia v1.6.1 following these steps:
```
wget https://julialang-s3.julialang.org/bin/linux/x64/1.6/julia-1.6.1-linux-x86_64.tar.gz
tar zxvf julia-1.6.1-linux-x86_64.tar.gz
```

In the julia command line I then put:
```
using Pkg
Pkg.add("Circuitscape")
Pkg.test("Circuitscape")
```

I also tried the master version of Circuitscape
```
Pkg.add(PackageSpec(name="Circuitscape", rev = "master"))
```

However I receive errors on some of the required tests which I think is stopping my next process from running correctly.
```
┌ Warning: 2021-06-28 11:53:14 : Code to include pairs is activated, some entries did not match with focal node file. Some focal nodes may have been dropped
└ @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/logging.jl:25
Test Summary:          | Pass  Error  Total
Double Precision Tests |  335      2    337
  Network Pairwise     |   93            93
  Network Advanced     |   12            12
  Raster Pairwise      |           1      1
  Raster Advanced      |           1      1
  Raster One to All    |  134           134
  Raster All to One    |   96            96
ERROR: LoadError: Some tests did not pass: 335 passed, 0 failed, 2 errored, 0 broken.
in expression starting at /pub28/eve/.julia/packages/Circuitscape/Qr6wW/test/runtests.jl:15
ERROR: Package Circuitscape errored during testing
Stacktrace:
  [1] pkgerror(msg::String)
    @ Pkg.Types /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:55
  [2] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing)
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1689
  [3] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, test_fn::Nothing, julia_args::Cmd, test_args::Cmd, kwargs::Base.Iterators.Pairs{Symbol, Base.TTY, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.TTY}}})
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:342
  [4] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.TTY, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:79
  [5] test(pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:77
  [6] #test#63
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:75 [inlined]
  [7] test
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:75 [inlined]
  [8] #test#62
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:74 [inlined]
  [9] test(pkg::String)
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:74
 [10] top-level scope
    @ REPL[7]:1
```

I am working on a server and I do not have sudo permissions, so mainly work with `conda`.

Please let me know how I can solve this error.
Thank you,
Eve

**Comment by EveTC** [comment]
Related issues - #188  and #178 although I am running it on a Linux server via Ubuntu not Windows and have additional errors,

**Comment by ranjanan** [comment]
I was able to reproduce this on one machine and not the other. It seems that Pardiso has not been built properly. Could you try running `using Pkg; Pkg.add("Pardiso"); Pkg.build("Pardiso")`? Run tests again after that. 

If it still fails, could you run the following code: 
```julia
using Pardiso, SparseArrays

ps = MKLPardisoSolver()

A = sparse(rand(10, 10))
B = rand(10, 2)
X = zeros(10, 2)
solve!(ps, X, A, B)
```
and see if that fails?

**Comment by EveTC** [comment]
Hi @ranjanan - thank you so much for your help.

Yes that looks exactly like my error, here is my full error output:
```
Raster Advanced: Error During Test at /pub28/eve/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:317
  Got exception outside of a @test
  could not load library "libmkl_rt"
  libmkl_rt.so: cannot open shared object file: No such file or directory
  Stacktrace:
    [1] ccall_pardiso(ps::Pardiso.MKLPardisoSolver, N::Int32, AA::Vector{Float64}, IA::Vector{Int32}, JA::Vector{Int32}, NRHS::Int32, B::Vector{Float64}, X::Vector{Float64})
      @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/mkl_pardiso.jl:57
    [2] pardiso(ps::Pardiso.MKLPardisoSolver, X::Vector{Float64}, A::SparseArrays.SparseMatrixCSC{Float64, Int64}, B::Vector{Float64})
      @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/Pardiso.jl:335
    [3] #_#34
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:448 [inlined]
    [4] MKLPardisoFactorize (repeats 2 times)
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:446 [inlined]
    [5] solve_linear_system(factor::Circuitscape.MKLPardisoFactorize, matrix::SparseArrays.SparseMatrixCSC{Float64, Int64}, rhs::Vector{Float64})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/core.jl:625
    [6] macro expansion
      @ ./timing.jl:287 [inlined]
    [7] multiple_solve(s::Circuitscape.MKLPardisoSolver, matrix::SparseArrays.SparseMatrixCSC{Float64, Int64}, sources::Vector{Float64}, suppress_info::Bool)
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:327
    [8] multiple_solver(cfg::Dict{String, String}, solver::Circuitscape.MKLPardisoSolver, a::SparseArrays.SparseMatrixCSC{Float64, Int64}, sources::Vector{Float64}, grounds::Vector{Float64}, finitegrounds::Vector{Float64})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:291
    [9] advanced_kernel(prob::Circuitscape.AdvancedProblem{Float64, Int64, Circuitscape.MKLPardisoSolver}, flags::Circuitscape.RasterFlags, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:206
   [10] raster_advanced(T::Type, V::Type, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:30
   [11] _compute(T::Type, V::Type, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/run.jl:45
   [12] macro expansion
      @ ./timing.jl:287 [inlined]
   [13] compute(path::String)
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/run.jl:31
   [14] macro expansion
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:321 [inlined]
   [15] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [16] macro expansion
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:319 [inlined]
   [17] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [18] runtests(f::typeof(compute))
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:266
   [19] top-level scope
      @ ~/.julia/packages/Circuitscape/Qr6wW/test/runtests.jl:16
   [20] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [21] top-level scope
      @ none:6
   [22] eval
      @ ./boot.jl:360 [inlined]
   [23] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:261
   [24] _start()
      @ Base ./client.jl:485
```

1. I then ran the below as suggested, see outputs below as well.
```
using Pkg
Pkg.add("Pardiso")

   Resolving package versions...
    Updating `~/.julia/environments/v1.6/Project.toml`
  [46dd5b70] + Pardiso v0.5.0
  No Changes to `~/.julia/environments/v1.6/Manifest.toml`

Pkg.build("Pardiso")

     Building Pardiso → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/8dbcb7c3c8066167825338732450db0ed1f2cad0/build.log`
     Precompiling project...
     1 dependency successfully precompiled in 2 seconds (60 already precompiled)

```

2. Run the test again `Pkg.test("Circuitscape")` - still receive what looks like the same error as above. Printed below incase there are and subtle changes
```
Raster Advanced: Error During Test at /pub28/eve/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:317
  Got exception outside of a @test
  could not load library "libmkl_rt"
  libmkl_rt.so: cannot open shared object file: No such file or directory
  Stacktrace:
    [1] ccall_pardiso(ps::Pardiso.MKLPardisoSolver, N::Int32, AA::Vector{Float64}, IA::Vector{Int32}, JA::Vector{Int32}, NRHS::Int32, B::Vector{Float64}, X::Vector{Float64})
      @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/mkl_pardiso.jl:57
    [2] pardiso(ps::Pardiso.MKLPardisoSolver, X::Vector{Float64}, A::SparseArrays.SparseMatrixCSC{Float64, Int64}, B::Vector{Float64})
      @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/Pardiso.jl:335
    [3] #_#34
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:448 [inlined]
    [4] MKLPardisoFactorize (repeats 2 times)
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:446 [inlined]
    [5] solve_linear_system(factor::Circuitscape.MKLPardisoFactorize, matrix::SparseArrays.SparseMatrixCSC{Float64, Int64}, rhs::Vector{Float64})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/core.jl:625
    [6] macro expansion
      @ ./timing.jl:287 [inlined]
    [7] multiple_solve(s::Circuitscape.MKLPardisoSolver, matrix::SparseArrays.SparseMatrixCSC{Float64, Int64}, sources::Vector{Float64}, suppress_info::Bool)
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:327
    [8] multiple_solver(cfg::Dict{String, String}, solver::Circuitscape.MKLPardisoSolver, a::SparseArrays.SparseMatrixCSC{Float64, Int64}, sources::Vector{Float64}, grounds::Vector{Float64}, finitegrounds::Vector{Float64})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:291
    [9] advanced_kernel(prob::Circuitscape.AdvancedProblem{Float64, Int64, Circuitscape.MKLPardisoSolver}, flags::Circuitscape.RasterFlags, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:206
   [10] raster_advanced(T::Type, V::Type, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/raster/advanced.jl:30
   [11] _compute(T::Type, V::Type, cfg::Dict{String, String})
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/run.jl:45
   [12] macro expansion
      @ ./timing.jl:287 [inlined]
   [13] compute(path::String)
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/run.jl:31
   [14] macro expansion
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:321 [inlined]
   [15] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [16] macro expansion
      @ ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:319 [inlined]
   [17] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [18] runtests(f::typeof(compute))
      @ Circuitscape ~/.julia/packages/Circuitscape/Qr6wW/src/utils.jl:266
   [19] top-level scope
      @ ~/.julia/packages/Circuitscape/Qr6wW/test/runtests.jl:16
   [20] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [21] top-level scope
      @ none:6
   [22] eval
      @ ./boot.jl:360 [inlined]
   [23] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:261
   [24] _start()
      @ Base ./client.jl:485
```

3. Run following code (output included):
```
using Pardiso, SparseArrays

ps = MKLPardisoSolver()
       
       MKLPardisoSolver:
        Matrix type: Real nonsymmetric
        Phase: Analysis, numerical factorization, solve, iterative refinement

A = sparse(rand(10, 10))    
B = rand(10, 2)
X = zeros(10, 2)

solve!(ps, X, A, B)

ERROR: could not load library "libmkl_rt"
libmkl_rt.so: cannot open shared object file: No such file or directory
Stacktrace:
 [1] ccall_pardisoinit(ps::MKLPardisoSolver)
   @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/mkl_pardiso.jl:47
 [2] pardisoinit
   @ ~/.julia/packages/Pardiso/GDj1P/src/Pardiso.jl:196 [inlined]
 [3] solve!(ps::MKLPardisoSolver, X::Matrix{Float64}, A::SparseMatrixCSC{Float64, Int64}, B::Matrix{Float64}, T::Symbol)
   @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/Pardiso.jl:269
 [4] solve!(ps::MKLPardisoSolver, X::Matrix{Float64}, A::SparseMatrixCSC{Float64, Int64}, B::Matrix{Float64})
   @ Pardiso ~/.julia/packages/Pardiso/GDj1P/src/Pardiso.jl:228
 [5] top-level scope
   @ REPL[13]:1
```


Ah ha!  I imagine I am missing the library "libmkl_rt"?

**Comment by ranjanan** [comment]
What does `Pardiso.libmkl_rt` return?

**Comment by EveTC** [comment]
`Pardiso.libmkl_rt` returns:
```
"libmkl_rt"
```

Ran:
```
using Pardiso
using Libdl
Libdl.dlopen(Pardiso.libmkl_rt)
```

Return:
```
ERROR: could not load library "libmkl_rt"
libmkl_rt.so: cannot open shared object file: No such file or directory
Stacktrace:
 [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
   @ Base.Libc.Libdl ./libdl.jl:114
 [2] dlopen (repeats 2 times)
   @ ./libdl.jl:114 [inlined]
 [3] top-level scope
   @ REPL[19]:1
```

**Comment by EveTC** [comment]
Is this thread relevant? https://github.com/JuliaSparse/Pardiso.jl/issues/69

**Comment by ranjanan** [comment]
I saw that thread, but I saw something similar on one Windows machine so it might not be Linux specific. 

As a final check, could you do run this too: 
```julia
Pardiso.MKL_jll.libmkl_rt
```
and 
```julia
using Pardiso
using Libdl
Libdl.dlopen(Pardiso.MKL_jll.libmkl_rt)
```
If that doesn't work, let's file an issue in Pardiso: I would create a separate directory, and run the following code to see if it reproduces the issue:  
```julia
using Pkg; Pkg.add("Pardiso"); 
using Pardiso, SparseArrays

ps = MKLPardisoSolver()

A = sparse(rand(10, 10))
B = rand(10, 2)
X = zeros(10, 2)
solve!(ps, X, A, B)
```
If that does, I would file an issue here: https://github.com/JuliaSparse/Pardiso.jl/issues with that reproducible code while mentioning the output of `versioninfo()` and the version of Pardiso.

**Comment by EveTC** [comment]
1. Ran `Pardiso.MKL_jll.libmkl_rt` output:
```
"/pub28/eve/.julia/artifacts/a8e009985328801a84c9af6610f94f77a7c12852/lib/libmkl_rt.so"
```

2. Ran `using Pardiso; using Libdl; Libdl.dlopen(Pardiso.MKL_jll.libmkl_rt)` output:
```
Ptr{Nothing} @0x0000000000d0c130
```

**Comment by EveTC** [comment]
What do you mean by "create a seperate directory"?

**Comment by EveTC** [comment]
This "/pub28/eve/.julia/artifacts/a8e009985328801a84c9af6610f94f77a7c12852/lib/libmkl_rt.so" slightly confuses me as my `julia` binaries are in a miniconda environment whichi is this: `/pub28/eve/bin/miniconda3/envs/Julia/julia-1.6.1/bin/julia`

**Comment by ranjanan** [comment]
Don't think that should be a problem. This is to confirm that the MKL binary (that Pardiso calls) has indeed been downloaded onto your computer. Does this snippet of code still fail: 
```julia
using Pardiso, SparseArrays

ps = MKLPardisoSolver()

A = sparse(rand(10, 10))
B = rand(10, 2)
X = zeros(10, 2)
solve!(ps, X, A, B)
```
?

**Comment by EveTC** [comment]
Yes same error as before :(

**Comment by ranjanan** [comment]
I don't think that would be necessary. I am going to post an issue on Pardiso.jl and check if anyone can help with it.

**Comment by EveTC** [comment]
Ok so I deleted `/pub28/eve/.julia`, strated a new conda environment and reinstalled `julia` using:

```
wget https://julialang-s3.julialang.org/bin/linux/x64/1.6/julia-1.6.1-linux-x86_64.tar.gz
tar zxvf julia-1.6.1-linux-x86_64.tar.gz
```

Then instead of first installing Circuitscape I ran the commands you provided me with for `Pardiso` as so:

```
using Pkg
Pkg.add("Pardiso")
Pkg.build("Pardiso")
```

I then ran your Parsdio test code and I no longer get an error but an output!
```
using Pardiso, SparseArrays

ps = MKLPardisoSolver()

A = sparse(rand(10, 10))
B = rand(10, 2)
X = zeros(10, 2)
solve!(ps, X, A, B)

10×2 Matrix{Float64}:
  0.471564   0.584395
 -0.590634  -0.528925
 -0.229152   1.10205
 -0.642087  -1.62416
  0.723133  -0.295956
  1.22932    1.51893
  0.299838   0.647085
 -0.215421   0.384913
  0.299716   0.0351154
 -0.456074  -0.50306
```

Ok - I will now try to install Circuitscape and test.....
```
using Pkg
Pkg.add("Circuitscape")
     Building TimeZones → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/81753f400872e5074768c9a77d4c44e70d409ef0/build.log`
     Precompiling project...
     46 dependencies successfully precompiled in 17 seconds (12 already precompiled)

Pkg.test("Circuitscape")

Test Summary:          | Pass  Total
Double Precision Tests |  662    662
     Testing Circuitscape tests passed
```

It worked! Hallelujah! Thank you so so much for all your help @ranjanan - I could not have done it without you 👍

**Comment by ranjanan** [comment]
Interesting. I'll add this to the issue.

**Comment by EveTC** [comment]
Yes not sure why the order seems important but Im just pleased to see it is working. Now to test with ResistanceGA 🤞.
Thank you again 😃

**Comment by vlandau** [comment]
@ranjanan are you keeping this open until the Pardiso issue you created is closed?

**Comment by ranjanan** [comment]
Yep. Once a new version of Pardiso is tagged I'll bump the versions up and then close this. Then I'll make a minor CS release.

**Comment by ranjanan** [comment]
Closed by #304
